File: resources/androidstudio/app/src/main/java/com/nativephp/mobile/ui/nativerender/TextRenderer.kt
package com.nativephp.mobile.ui.nativerender

import android.content.Context
import android.content.res.Configuration
import android.graphics.Typeface
import android.text.TextUtils
import android.view.Gravity
import android.view.View
import android.widget.TextView

class TextViewRenderer : NativeViewRenderer {
    override fun createView(context: Context, node: NativeUINode): View {
        val tv = TextView(context)
        tv.ellipsize = TextUtils.TruncateAt.END
        applyProps(tv, node)
        return tv
    }

    override fun updateView(view: View, node: NativeUINode) {
        val tv = view as? TextView ?: return
        applyProps(tv, node)
    }

    private fun applyProps(tv: TextView, node: NativeUINode) {
        val p = node.props
        tv.text = p.getString("text")

        val fontSize = p.getFloat("font_size", 16f)
        tv.textSize = fontSize

        val dark = (tv.context.resources.configuration.uiMode and
            Configuration.UI_MODE_NIGHT_MASK) == Configuration.UI_MODE_NIGHT_YES
        val darkColor = if (dark) p.getColor("dark_color", 0) else 0
        val color = if (darkColor != 0) darkColor else p.getColor("color", 0xFF000000.toInt())
        tv.setTextColor(color)

        val weight = p.getInt("font_weight")
        tv.typeface = Typeface.create(Typeface.DEFAULT, resolveTypefaceStyle(weight))

        tv.gravity = resolveGravity(p.getInt("text_align")) or Gravity.CENTER_VERTICAL

        val maxLines = p.getInt("max_lines")
        tv.maxLines = if (maxLines > 0) maxLines else Int.MAX_VALUE
    }
}

private fun resolveTypefaceStyle(weight: Int): Int {
    return when (weight) {
        1, 2, 3 -> Typeface.NORMAL  // thin, light, normal
        4, 5 -> Typeface.BOLD       // medium, semibold (closest match)
        6, 7 -> Typeface.BOLD       // bold, extrabold
        else -> Typeface.NORMAL
    }
}

private fun resolveGravity(align: Int): Int {
    return when (align) {
        0 -> Gravity.START
        1 -> Gravity.CENTER
        2 -> Gravity.END
        else -> Gravity.START
    }
}

File: resources/xcode/NativePHP/NativeRender/TextRenderer.swift
import UIKit

struct TextViewRenderer: NativeViewRenderer {
    func createView(node: NativeUINode) -> UIView {
        let label = UILabel()
        label.numberOfLines = 0
        applyProps(label, node: node)
        return label
    }

    func updateView(_ view: UIView, node: NativeUINode) {
        guard let label = view as? UILabel else { return }
        applyProps(label, node: node)
    }

    private func applyProps(_ label: UILabel, node: NativeUINode) {
        let p = node.props
        label.text = p.getString("text")
        label.font = .systemFont(
            ofSize: CGFloat(p.getFloat("font_size", default: 16)),
            weight: resolveWeight(p.getInt("font_weight"))
        )
        let dark: Bool = {
            if #available(iOS 13.0, *) {
                // Use label's own traitCollection for reliable dark mode detection
                return label.traitCollection.userInterfaceStyle == .dark
            }
            return false
        }()
        if p.has("dark_color") {
            print("DARK DEBUG TEXT: isDark=\(dark) has_dark_color=true dark_color_val=\(p.getColor("dark_color", default: 0)) text=\(p.getString("text").prefix(20))")
        }
        let darkColor = dark ? p.getColor("dark_color", default: 0) : 0
        let textArgb = darkColor != 0 ? darkColor : p.getColor("color", default: 0xFF000000)
        label.textColor = UIColor(argb: textArgb)
        label.textAlignment = resolveAlignment(p.getInt("text_align"))
        let maxLines = p.getInt("max_lines")
        label.numberOfLines = maxLines > 0 ? maxLines : 0
        label.lineBreakMode = .byTruncatingTail
        label.adjustsFontSizeToFitWidth = true
        label.minimumScaleFactor = 0.5
    }
}

private func resolveWeight(_ weight: Int) -> UIFont.Weight {
    switch weight {
    case 1: return .thin
    case 2: return .light
    case 3: return .regular
    case 4: return .medium
    case 5: return .semibold
    case 6: return .bold
    case 7: return .heavy
    default: return .regular
    }
}

private func resolveAlignment(_ align: Int) -> NSTextAlignment {
    switch align {
    case 0: return .left
    case 1: return .center
    case 2: return .right
    default: return .natural
    }
}
